File: backend/app/api/conversations.py
from fastapi import APIRouter, HTTPException, Depends, status
from pydantic import BaseModel
from typing import List, Optional, Dict, Any
from app.services.conversation_service import conversation_service
from app.services.ai_service import ai_service
from app.core.websocket_manager import websocket_manager
from app.api.auth import get_current_user

router = APIRouter(prefix="/conversations", tags=["conversations"])

class ConversationCreate(BaseModel):
    project_id: str
    title: Optional[str] = None

class MessageCreate(BaseModel):
    content: str
    message_type: str = "user"  # user, ai, system

class ConversationResponse(BaseModel):
    id: str
    user_id: str
    project_id: str
    title: str
    messages: List[Dict[str, Any]]
    context: Dict[str, Any]
    status: str
    created_at: str
    updated_at: str

@router.post("/", response_model=ConversationResponse)
async def create_conversation(
    conversation_data: ConversationCreate,
    current_user: dict = Depends(get_current_user)
):
    """Create a new conversation"""
    try:
        conversation = await conversation_service.create_conversation(
            user_id=current_user["id"],
            project_id=conversation_data.project_id,
            title=conversation_data.title
        )
        
        return ConversationResponse(**conversation)
    
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e)
        )

@router.get("/", response_model=List[ConversationResponse])
async def get_user_conversations(
    project_id: Optional[str] = None,
    current_user: dict = Depends(get_current_user)
):
    """Get conversations for the current user"""
    conversations = await conversation_service.get_user_conversations(
        user_id=current_user["id"],
        project_id=project_id
    )
    
    return [ConversationResponse(**conv) for conv in conversations]

@router.post("/{conversation_id}/messages")
async def add_message(
    conversation_id: str,
    message_data: MessageCreate,
    current_user: dict = Depends(get_current_user)
):
    """Add a message to conversation and get AI response"""
    # Add user message
    user_message = {
        "content": message_data.content,
        "type": message_data.message_type,
        "user_id": current_user["id"]
    }
    
    success = await conversation_service.add_message(conversation_id, user_message)
    if not success:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Conversation not found"
        )
    
    # Generate AI response if user message
    if message_data.message_type == "user":
        # Analyze user intent
        intent_analysis = await ai_service.analyze_user_intent(
            message=message_data.content,
            context={"conversation_id": conversation_id}
        )
        
        # Generate AI response based on intent
        if intent_analysis["intent"] == "create":
            ai_response = await ai_service.generate_wireframe_suggestions(
                description=message_data.content
            )
            response_content = f"I can help you create that! Here are some wireframe suggestions: {ai_response}"
        else:
            response_content = "I understand your request. How can I help you with your wireframe design?"
        
        # Add AI response message
        ai_message = {
            "content": response_content,
            "type": "ai",
            "intent_analysis": intent_analysis
        }
        
        await conversation_service.add_message(conversation_id, ai_message)
        
        # Broadcast AI response via WebSocket
        await websocket_manager.broadcast_ai_response(
            current_user["id"],
            {"message": ai_message, "conversation_id": conversation_id}
        )
    
    return {"message": "Message added successfully"}

@router.get("/{conversation_id}/messages")
async def get_conversation_messages(
    conversation_id: str,
    current_user: dict = Depends(get_current_user)
):
    """Get conversation message history"""
    messages = await conversation_service.get_conversation_history(conversation_id)
    if messages is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Conversation not found"
        )
    
    return {"messages": messages}

@router.put("/{conversation_id}/context")
async def update_conversation_context(
    conversation_id: str,
    context_updates: Dict[str, Any],
    current_user: dict = Depends(get_current_user)
):
    """Update conversation context"""
    success = await conversation_service.update_context(conversation_id, context_updates)
    if not success:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Conversation not found"
        )
    
    return {"message": "Context updated successfully"}

@router.post("/{conversation_id}/briefs")
async def create_context_brief(
    conversation_id: str,
    brief_type: str,
    content: str,
    metadata: Optional[Dict[str, Any]] = None,
    current_user: dict = Depends(get_current_user)
):
    """Create a context brief for AI memory"""
    brief = await conversation_service.create_context_brief(
        conversation_id=conversation_id,
        brief_type=brief_type,
        content=content,
        metadata=metadata or {}
    )
    
    return {"brief": brief}

@router.get("/{conversation_id}/briefs")
async def get_context_briefs(
    conversation_id: str,
    brief_type: Optional[str] = None,
    current_user: dict = Depends(get_current_user)
):
    """Get context briefs for a conversation"""
    briefs = await conversation_service.get_context_briefs(
        conversation_id=conversation_id,
        brief_type=brief_type
    )
    
    return {"briefs": briefs}

File: backend-js/src/routes/conversations.js
import { Router } from 'express';
import { body, param, validationResult } from 'express-validator';
import { db } from '../config/database';
import { aiService } from '../services/aiService';
import { authenticateToken } from '../middleware/auth';
import { logger } from '../utils/logger';
import { Conversation } from '../types';

const router = Router();

// Validation rules
const createConversationValidation = [
  body('projectId').isUUID().withMessage('Invalid project ID'),
  body('messageContent').isObject().withMessage('Message content is required'),
];

const conversationIdValidation = [
  param('id').isUUID().withMessage('Invalid conversation ID'),
];

// Create conversation message
router.post('/', authenticateToken, createConversationValidation, async (req, res) => {
  try {
    const errors = validationResult(req);
    if (!errors.isEmpty()) {
      return res.status(400).json({
        success: false,
        error: 'Validation failed',
        details: errors.array(),
        timestamp: new Date()
      });
    }

    const { projectId, messageContent } = req.body;

    // Verify project ownership
    const project = await db.findById('projects', projectId);
    if (!project || project.user_id !== req.user!.userId) {
      return res.status(404).json({
        success: false,
        error: 'Project not found',
        timestamp: new Date()
      });
    }

    // Analyze user intent using AI service
    let intentAnalysis = null;
    let aiResponse = null;
    let generatedWireframeId = null;

    try {
      // Get project context for better AI understanding
      const projectContext = {
        projectType: project.project_type,
        targetPlatform: project.target_platform,
        industry: project.industry,
        targetAudience: project.target_audience,
        aiContext: project.ai_context
      };

      // Analyze user intent
      intentAnalysis = await aiService.analyzeUserIntent(
        messageContent.text || messageContent.message || '',
        projectContext
      );

      // Generate AI response based on intent
      if (intentAnalysis.intent === 'create' || intentAnalysis.intent === 'wireframe') {
        // Generate wireframe suggestions
        const wireframeSuggestions = await aiService.generateWireframeSuggestions(
          messageContent.text || messageContent.message || '',
          projectContext
        );

        aiResponse = {
          type: 'wireframe_suggestions',
          suggestions: wireframeSuggestions,
          message: 'I\'ve generated some wireframe suggestions based on your request.',
          confidence: intentAnalysis.confidence || 0.8
        };

        // Auto-create wireframe if confidence is high
        if (intentAnalysis.confidence > 0.8 && wireframeSuggestions.length > 0) {
          const wireframeData = {
            project_id: projectId,
            screen_name: intentAnalysis.screenName || 'AI Generated Screen',
            screen_type: intentAnalysis.screenType || 'page',
            device_type: intentAnalysis.deviceType || 'desktop',
            content_structure: wireframeSuggestions[0].components || {},
            design_system: wireframeSuggestions[0].styling || {},
            ai_reasoning: wireframeSuggestions[0],
            generation_prompt: messageContent.text || messageContent.message,
            version_number: 1,
            is_active: true,
            status: 'draft'
          };

          const wireframe = await db.create('wireframes', wireframeData);
          if (wireframe) {
            generatedWireframeId = wireframe.id;
            aiResponse.generatedWireframeId = wireframe.id;
          }
        }
      } else if (intentAnalysis.intent === 'component') {
        // Generate component suggestions
        const componentSuggestions = await aiService.generateUIComponents(
          intentAnalysis.componentType || 'button',
          { requirements: messageContent.text || messageContent.message }
        );

        aiResponse = {
          type: 'component_suggestions',
          suggestions: componentSuggestions,
          message: 'Here are some component suggestions for your request.',
          confidence: intentAnalysis.confidence || 0.7
        };
      } else {
        // General AI response
        aiResponse = {
          type: 'general_response',
          message: 'I understand your request. How would you like me to help you with your wireframe?',
          suggestions: [],
          confidence: intentAnalysis.confidence || 0.5
        };
      }
    } catch (error) {
      logger.warn('AI processing failed, using fallback response:', error);
      aiResponse = {
        type: 'fallback_response',
        message: 'I received your message. Could you provide more details about what you\'d like to create?',
        confidence: 0.3
      };
    }

    // Save conversation
    const conversationData = {
      project_id: projectId,
      user_id: req.user!.userId,
      message_content: messageContent,
      ai_response: aiResponse,
      intent_analysis: intentAnalysis,
      context_used: project.ai_context || {},
      generated_wireframe_id: generatedWireframeId
    };

    const conversation = await db.create<Conversation>('conversations', conversationData);
    if (!conversation) {
      throw new Error('Failed to create conversation');
    }

    res.status(201).json({
      success: true,
      data: { 
        conversation,
        aiResponse,
        generatedWireframeId
      },
      message: 'Message processed successfully',
      timestamp: new Date()
    });
  } catch (error: any) {
    logger.error('Create conversation error:', error);
    res.status(500).json({
      success: false,
      error: error.message || 'Failed to process message',
      timestamp: new Date()
    });
  }
});

// Get project conversations
router.get('/project/:projectId', authenticateToken, async (req, res) => {
  try {
    const { projectId } = req.params;

    // Verify project ownership
    const project = await db.findById('projects', projectId);
    if (!project || project.user_id !== req.user!.userId) {
      return res.status(404).json({
        success: false,
        error: 'Project not found',
        timestamp: new Date()
      });
    }

    const conversations = await db.getProjectConversations(projectId);

    res.json({
      success: true,
      data: { conversations },
      timestamp: new Date()
    });
  } catch (error: any) {
    logger.error('Get project conversations error:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to get conversations',
      timestamp: new Date()
    });
  }
});

// Get conversation by ID
router.get('/:id', authenticateToken, conversationIdValidation, async (req, res) => {
  try {
    const errors = validationResult(req);
    if (!errors.isEmpty()) {
      return res.status(400).json({
        success: false,
        error: 'Validation failed',
        details: errors.array(),
        timestamp: new Date()
      });
    }

    const conversation = await db.findById<Conversation>('conversations', req.params.id);
    if (!conversation) {
      return res.status(404).json({
        success: false,
        error: 'Conversation not found',
        timestamp: new Date()
      });
    }

    // Verify project ownership
    const project = await db.findById('projects', conversation.project_id);
    if (!project || project.user_id !== req.user!.userId) {
      return res.status(404).json({
        success: false,
        error: 'Conversation not found',
        timestamp: new Date()
      });
    }

    res.json({
      success: true,
      data: { conversation },
      timestamp: new Date()
    });
  } catch (error: any) {
    logger.error('Get conversation error:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to get conversation',
      timestamp: new Date()
    });
  }
});

// Delete conversation
router.delete('/:id', authenticateToken, conversationIdValidation, async (req, res) => {
  try {
    const errors = validationResult(req);
    if (!errors.isEmpty()) {
      return res.status(400).json({
        success: false,
        error: 'Validation failed',
        details: errors.array(),
        timestamp: new Date()
      });
    }

    const conversation = await db.findById<Conversation>('conversations', req.params.id);
    if (!conversation) {
      return res.status(404).json({
        success: false,
        error: 'Conversation not found',
        timestamp: new Date()
      });
    }

    // Verify project ownership
    const project = await db.findById('projects', conversation.project_id);
    if (!project || project.user_id !== req.user!.userId) {
      return res.status(404).json({
        success: false,
        error: 'Conversation not found',
        timestamp: new Date()
      });
    }

    await db.delete('conversations', req.params.id);

    res.json({
      success: true,
      message: 'Conversation deleted successfully',
      timestamp: new Date()
    });
  } catch (error: any) {
    logger.error('Delete conversation error:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to delete conversation',
      timestamp: new Date()
    });
  }
});

// Get conversation context for AI
router.get('/project/:projectId/context', authenticateToken, async (req, res) => {
  try {
    const { projectId } = req.params;
    const { limit = 10 } = req.query;

    // Verify project ownership
    const project = await db.findById('projects', projectId);
    if (!project || project.user_id !== req.user!.userId) {
      return res.status(404).json({
        success: false,
        error: 'Project not found',
        timestamp: new Date()
      });
    }

    const conversations = await db.findMany<Conversation>(
      'conversations',
      { project_id: projectId },
      { 
        limit: parseInt(limit as string), 
        orderBy: 'created_at', 
        ascending: false 
      }
    );

    // Extract context from recent conversations
    const context = {
      recentMessages: conversations.map(c => ({
        userMessage: c.message_content,
        aiResponse: c.ai_response,
        intent: c.intent_analysis,
        timestamp: c.created_at
      })),
      projectContext: {
        type: project.project_type,
        platform: project.target_platform,
        industry: project.industry,
        audience: project.target_audience
      }
    };

    res.json({
      success: true,
      data: { context },
      timestamp: new Date()
    });
  } catch (error: any) {
    logger.error('Get conversation context error:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to get conversation context',
      timestamp: new Date()
    });
  }
});

// Regenerate AI response
router.post('/:id/regenerate', authenticateToken, conversationIdValidation, async (req, res) => {
  try {
    const errors = validationResult(req);
    if (!errors.isEmpty()) {
      return res.status(400).json({
        success: false,
        error: 'Validation failed',
        details: errors.array(),
        timestamp: new Date()
      });
    }

    const conversation = await db.findById<Conversation>('conversations', req.params.id);
    if (!conversation) {
      return res.status(404).json({
        success: false,
        error: 'Conversation not found',
        timestamp: new Date()
      });
    }

    // Verify project ownership
    const project = await db.findById('projects', conversation.project_id);
    if (!project || project.user_id !== req.user!.userId) {
      return res.status(404).json({
        success: false,
        error: 'Conversation not found',
        timestamp: new Date()
      });
    }

    // Regenerate AI response
    const projectContext = {
      projectType: project.project_type,
      targetPlatform: project.target_platform,
      industry: project.industry,
      targetAudience: project.target_audience,
      aiContext: project.ai_context
    };

    const newIntentAnalysis = await aiService.analyzeUserIntent(
      conversation.message_content.text || conversation.message_content.message || '',
      projectContext
    );

    let newAiResponse;
    if (newIntentAnalysis.intent === 'create' || newIntentAnalysis.intent === 'wireframe') {
      const wireframeSuggestions = await aiService.generateWireframeSuggestions(
        conversation.message_content.text || conversation.message_content.message || '',
        projectContext
      );

      newAiResponse = {
        type: 'wireframe_suggestions',
        suggestions: wireframeSuggestions,
        message: 'Here are some updated wireframe suggestions.',
        confidence: newIntentAnalysis.confidence || 0.8
      };
    } else {
      newAiResponse = {
        type: 'general_response',
        message: 'Let me provide you with a different perspective on your request.',
        confidence: newIntentAnalysis.confidence || 0.5
      };
    }

    // Update conversation
    const updatedConversation = await db.update<Conversation>('conversations', req.params.id, {
      ai_response: newAiResponse,
      intent_analysis: newIntentAnalysis
    });

    res.json({
      success: true,
      data: { 
        conversation: updatedConversation,
        aiResponse: newAiResponse
      },
      message: 'AI response regenerated successfully',
      timestamp: new Date()
    });
  } catch (error: any) {
    logger.error('Regenerate AI response error:', error);
    res.status(500).json({
      success: false,
      error: error.message || 'Failed to regenerate response',
      timestamp: new Date()
    });
  }
});

export default router;
